<a href="https://colab.research.google.com/github/Simranpreet-Singh-TIET/Dataset_Movies/blob/main/ML_movie_Recommendation_System.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

A movie recommendation system recommends similar movies on the basis of the movie you just entered


**AI Project** by **Harsimran Singh Dhillon** [ 102003476 ] & **Simranpreet Singh** [ 102003485 ]

# python code for the recommendaton system

In [ ]:
#importing libraries

import numpy as np
import pandas as pd
import difflib
from sklearn.feature_extraction.text import TfidfVectorizer# converts textual data into numbers
from sklearn.metrics.pairwise import cosine_similarity # implements cosine_similarity


Data Collection and Pre-processing

In [ ]:
movies_data = pd.read_csv('https://raw.githubusercontent.com/Simranpreet-Singh-TIET/Dataset_Movies/main/movies.csv')

In [ ]:
movies_data.head() # in order to get the overview of the data

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

In [ ]:
# finding the size of the dataframe
movies_data.shape

(4803, 24)

In [ ]:
# selecting the relevant features for selection aka feature selection
selected_features = ['genres', 'keywords', 'tagline', 'cast', 'director']
print(selected_features)

['genres', 'keywords', 'tagline', 'cast', 'director']


Data preparation

In [ ]:
# replacing the missing values with NULL string
for feature in selected_features:
  movies_data[feature] = movies_data[feature].fillna('')

In [ ]:
# combining all the five selected features
combined_features = movies_data['genres'] + movies_data['keywords'] + movies_data['tagline'] + movies_data['cast'] + movies_data['director']

In [ ]:
vectorizer = TfidfVectorizer()

In [ ]:
feature_vectors = vectorizer.fit_transform(combined_features)

In [ ]:
print(feature_vectors)

  (0, 3583)	0.16941894714909375
  (0, 20729)	0.2738578969459924
  (0, 16614)	0.15696073877453268
  (0, 14550)	0.2236809363648068
  (0, 23267)	0.16128139780622516
  (0, 26547)	0.19638671147741732
  (0, 22464)	0.20207089568711745
  (0, 21487)	0.21550128478931546
  (0, 27540)	0.19771357974524176
  (0, 27182)	0.23480088356130552
  (0, 21521)	0.1562356665943368
  (0, 18558)	0.2612168335104853
  (0, 18164)	0.08690831799482265
  (0, 27161)	0.1261988657937151
  (0, 24197)	0.07518543993419265
  (0, 22778)	0.2738578969459924
  (0, 4676)	0.2452909770424961
  (0, 26334)	0.13059723071916296
  (0, 22916)	0.33668756406929184
  (0, 10193)	0.16532432420812987
  (0, 4456)	0.21799675215510664
  (0, 9304)	0.2738578969459924
  (0, 21836)	0.09966592997173944
  (0, 8936)	0.11806131645084653
  (0, 444)	0.09109355212252294
  :	:
  (4801, 5449)	0.31510541707008866
  (4801, 22170)	0.31510541707008866
  (4801, 27361)	0.31510541707008866
  (4801, 27513)	0.3005604008026175
  (4801, 8121)	0.2611505149930138
  (4801,

Cosine Similarity

In [ ]:
# getting similarity scores using cosine similarity

similarity = cosine_similarity(feature_vectors)

In [ ]:
print(similarity)
similarity.shape

[[1.         0.06865296 0.01492221 ... 0.         0.         0.        ]
 [0.06865296 1.         0.02799128 ... 0.01243107 0.         0.        ]
 [0.01492221 0.02799128 1.         ... 0.         0.         0.        ]
 ...
 [0.         0.01243107 0.         ... 1.         0.         0.        ]
 [0.         0.         0.         ... 0.         1.         0.        ]
 [0.         0.         0.         ... 0.         0.         1.        ]]


(4803, 4803)

In [ ]:
# getting the movie name from the user
movie_name = input(' Enter the name of your favorite movie : ')


 Enter the name of your favorite movie interstellar


In [ ]:
#creating a list of all the movie names given in the data set
list_of_all_titles = movies_data['title'].tolist()
print(list_of_all_titles)

['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre', 'The Dark Knight Rises', 'John Carter', 'Spider-Man 3', 'Tangled', 'Avengers: Age of Ultron', 'Harry Potter and the Half-Blood Prince', 'Batman v Superman: Dawn of Justice', 'Superman Returns', 'Quantum of Solace', "Pirates of the Caribbean: Dead Man's Chest", 'The Lone Ranger', 'Man of Steel', 'The Chronicles of Narnia: Prince Caspian', 'The Avengers', 'Pirates of the Caribbean: On Stranger Tides', 'Men in Black 3', 'The Hobbit: The Battle of the Five Armies', 'The Amazing Spider-Man', 'Robin Hood', 'The Hobbit: The Desolation of Smaug', 'The Golden Compass', 'King Kong', 'Titanic', 'Captain America: Civil War', 'Battleship', 'Jurassic World', 'Skyfall', 'Spider-Man 2', 'Iron Man 3', 'Alice in Wonderland', 'X-Men: The Last Stand', 'Monsters University', 'Transformers: Revenge of the Fallen', 'Transformers: Age of Extinction', 'Oz: The Great and Powerful', 'The Amazing Spider-Man 2', 'TRON: Legacy', 'Cars 2', 'Green Lant

In [ ]:
# finding the close match for the movie name given by the user
find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)
print(find_close_match)

['Interstellar', 'Cinderella', 'Cinderella Man']


In [ ]:
close_match = find_close_match[0]

In [ ]:
# finding the index of the movie
index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]
print(index_of_the_movie)

95


In [ ]:
# getting a list of similar movies

similarity_score = list(enumerate(similarity[index_of_the_movie]))

In [ ]:
# sorting the movies based in their similairty score
sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True)
print(sorted_similar_movies)

[(95, 0.9999999999999998), (634, 0.18810182136681977), (3, 0.17227945118926624), (3688, 0.16776850046291833), (125, 0.16466211230592237), (123, 0.16291518799069465), (3439, 0.1577953189813501), (1446, 0.14551932044368782), (3645, 0.14459537419037097), (93, 0.13388761628828688), (3413, 0.12496493307019613), (2817, 0.1244156039648103), (2677, 0.12320700461930226), (270, 0.1208062477299534), (1852, 0.11824888665099632), (1645, 0.11691776209346612), (1018, 0.11668878300406381), (1889, 0.11319831339693032), (2285, 0.11181608060317516), (4526, 0.10913855261367603), (108, 0.10900020211407155), (1296, 0.10873473688966317), (471, 0.10851413065418086), (2366, 0.10563796686069471), (1196, 0.10561465836430597), (4083, 0.10511724685692819), (4328, 0.10504838059287593), (2790, 0.10410014465191086), (2524, 0.10346032498902429), (43, 0.10223812852652844), (250, 0.10222493053437272), (437, 0.10168751892183381), (1578, 0.10139618885361601), (119, 0.1011957967549897), (116, 0.10089211678973868), (1264, 0

In [ ]:
# print the name of similar movies based on the index
print('Movies suggested for you : \n')
i = 1

for movie in sorted_similar_movies:
  index = movie[0]
  title_from_index = movies_data[movies_data.index == index]['title'].values[0]
  if ( i <= 15 ):
    print(i, '.', title_from_index)
    i+= 1

Movies suggested for you : 

1 . Interstellar
2 . The Matrix
3 . The Dark Knight Rises
4 . Dear Frankie
5 . The Matrix Reloaded
6 . The Matrix Revolutions
7 . The Terminator
8 . The Tree of Life
9 . Gandhi, My Father
10 . Terminator 3: Rise of the Machines
11 . House at the End of the Street
12 . WarGames
13 . Good Deeds
14 . The Martian
15 . The Last Emperor


#Movie Recommendation System

In [ ]:
movie_name = input(' Enter the name of your favorite movie : ')
list_of_all_titles = movies_data['title'].tolist()
find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)
close_match = find_close_match[0]
index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]
similarity_score = list(enumerate(similarity[index_of_the_movie]))
sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True)
print('Movies suggested for you : \n')
i = 1

for movie in sorted_similar_movies:
  index = movie[0]
  title_from_index = movies_data[movies_data.index == index]['title'].values[0]
  if ( i <= 10 ):
    print(i, '.', title_from_index)
    i+= 1

 Enter the name of your favorite movie matrix
Movies suggested for you : 

1 . The Matrix
2 . The Matrix Reloaded
3 . The Matrix Revolutions
4 . Terminator 3: Rise of the Machines
5 . I, Robot
6 . Interstellar
7 . The Terminator
8 . 2001: A Space Odyssey
9 . Terminator Genisys
10 . Captain America: The First Avenger


# Limitations

this implementation of the Movie recommendor system only uses 5 categories hence, may not provide accurate result.
In order to improve accuracy, the number of fields should be increases

this implementation is content based and does not consider user behaviour or watch patterns

the size of the dataset is limited to 4803 movies and 24 columns of imformation about them, so the system would not work for any movie outside of the dataset.